src/ndk.js

```diff
@@ -1,7 +1,108 @@
-// https://github.com/appcelerator/titanium_mobile/blob/master/node_modules/titanium-sdk/lib/android.js
+import fs from 'fs';
+import path from 'path';
+import which from 'which';
 
-// toolchain, architecture stuff:
-// https://github.com/appcelerator/androidlib/blob/master/lib/env.js
+import * as util from './util';
+
+const cmd = util.cmd;
+const ndkBuild = `ndk-build${cmd}`;
+const ndkGdb = `ndk-gdb${cmd}`;
+let cache = null;
+
+/**
+ * Detects installed Android NDK.
+ *
+ * @param {Object} [opts] - An object with various params.
+ * @param {Boolean} [opts.bypassCache=false] - When true, forces scan for all Paths.
+ * @param {String} [opts.ndkPath] - Path to a known Android NDK directory.
+ * @returns {Promise}
+ */
+export function detect(opts = {}) {
+	if (cache && !opts.bypassCache) {
+		return Promise.resolve(cache);
+	}
+
+	const results = cache = {
+		ndk: {}
+	};
 
-export function detect() {
+	let ndkDir = opts.ndkPath || process.env.ANDROID_NDK;
+	const ndkPaths = [];
+	const searchDirs = util.getSearchPaths();
+
+	if (ndkDir) {
+		ndkDir = util.expandPath(ndkDir);
+		if (util.existsSync(ndkDir)) {
+			ndkPaths.push(ndkDir);
+		}
+	}
+
+	searchDirs
+		.map(dir => {
+			dir = util.expandPath(dir);
+			if (util.existsSync(dir)) {
+				fs.readdirSync(dir).forEach(sub => {
+					ndkPaths.push(path.join(dir, sub));
+				});
+			}
+		});
+
+	return Promise
+		.all(ndkPaths.map(p => isNDK(p)))
+		.then(values => results.ndk = values.filter(a => { return a; }).shift())
```

Comment:
We should return all valid NDKs, not just the first one.

Also, you can do `.filter(a => a)`.